File: Java/Algorithms/Sort/SelectionSort/SelectionSort.java
/**
 * SelectionSort 选择排序
 */
public class SelectionSort {

  private SelectionSort() {
  }

  public static void sort(int[] arr) {
    int n = arr.length;

    for (int i = 0; i < n; i++) {
      
      // 寻找 [i..n-1] 区间里的最小值的索引
      int minIndex = i;

      for (int j = i + 1; j < n; j++) {
        if (arr[j] < arr[minIndex]) {
          minIndex = j;
        }
      }
      swap(arr, i, minIndex);
    }
  }

  private static void swap(int[] arr, int i, int j) {
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
  }

  public static void main(String[] args) {
    int[] arr = { 10, 9, 8, 7, 6, 5, 4, 3, 2, 1 };
    SelectionSort.sort(arr);
    for (int i = 0; i < arr.length; i++) {
      System.out.println(arr[i]);
    }
    System.out.println();
  }

}

File: algorithms/sorting/selection-sort.ts
// 选择排序

// 算法步骤
// 1. 在未排序序列中找到最小元素，存放到排序序列的起始位置
// 2. 从剩余未排序元素中继续寻找最小元素，然后放到已排序序列的末尾。
// 3. 重复第二步，直到所有元素均排序完毕。

import { swap, defaultCompare, Compare } from "../util";

export function selectionSort<T>(array: T[], compareFn = defaultCompare): T[] {
  const len = array.length;
  let minIndex: number;

  for (let i = 0; i < len - 1; i++) {
    minIndex = i; // 寻找最小值索引，初始值为 i

    for (let j = i + 1; j < len; j++) {
      if (compareFn(array[minIndex], array[j]) === Compare.BIGGER_THAN) {
        minIndex = j; // 更新最小值索引
      }
    }

    if (i !== minIndex) swap(array, i, minIndex);
  }
  return array;
}
